### PR TITLE
Utilise file locking so we can track crashed processes

### PR DESCRIPTION
**Pull request by willmot**
We are currently [unable to tell whether the background PHP process that the backup is running in has crashed](https://github.com/humanmade/backupwordpress/issues/908) or is still running.  We create a `.backup_running` file to track that a backup is in progress, this file is then deleted when the backup completes. The problem is that if the backup process fatal errors, crashes or is killed for some reason that file is not deleted and thus the backup looks like it's still running.

This PR solves this issue with [file locking], which is supported in PHP with the [`flock`](http://php.net/manual/en/function.flock.php) function. Once set, a file lock ensures that the file cannot then be written to by another thread until the lock is removed by the first thread. The kicker for us, file locks are automatically cleaned up when a PHP process ends, no matter how it's ended. We can use that fact to detect crashed backups.

If the `.backup_running` file exists and is locked then the backup must still be running and crucially if the `.backup_running` file exists but the isn't locked then the backup process must have started and then ended before it could properly complete. Huzzah!

We're using [`symfony/LockHandler`](http://symfony.com/doc/current/components/filesystem/lock_handler.html) to make working with file locks easier.

### CODE REVIEW

**Comment by willmot** [comment]
This still needs more testing and ideally some robust unit tests before it's merged. Unfortunately unit testing is complex as it requires starting backups in multiple threads and then killing them. Work continues.

**Comment by willmot** [comment]
This would fix https://github.com/humanmade/backupwordpress/issues/908 and will help with https://github.com/humanmade/backupwordpress/issues/763, https://github.com/humanmade/backupwordpress/issues/762 & https://github.com/humanmade/backupwordpress/issues/752

**Comment by willmot** [comment]
Added in detection of crashed backups, we now cleanup properly and show a message:

screen shot 2016-03-12 at 01 10 01

#### Still Todo

- [ ] Mixup the unit tests
- [ ] Better error message
- [ ] Persistent notices aren't clearing up properly
- [ ] Test on Windows
- [ ] Check on a system which doesn't support file locking, we shouldn't default to `has_crashed`

**Comment by willmot** [comment]
https://github.com/humanmade/backupwordpress/commit/abb7e410a16bbf0fb25de477ec79896fedbf5b48 fixes https://github.com/humanmade/backupwordpress/issues/763

**Comment by willmot** [comment]
@pdewouters we should get a beta of this release going so Kat can test with people who are running into the issues this should solve.

**Comment by katmoody** [comment]
Consider adding one further item above @willmot and/or @pdewouters -- sending an email when this comes up? Or will that automatically send as per the email settings? I get a lot of notes from folks who never got a warning anything was wrong on their backups and don't check their admin regularly, so I'm thinking the email warning would be an important aspect.

**Comment by willmot** [comment]
@katmoody agreed that would be a very useful enhancement, we actually have that covered in #767

**Comment by willmot** [comment]
> Fixup the unit tests

I think unit testing with multiple PHP threads is going to be a blocker here. For now the unit tests are `markSkipped`.

**Comment by willmot** [comment]
On systems which don't support file locks there are 2 error conditions that will in practise mean backups will stop working for those users.

1. Locking never works: backups won't be able to start as the status file won't be locked in `Backup_Status::start()`.

2. If the status file is locked when starting the backup but checking the lock fails for some reason then the backup will be (incorrectly) marked as crashed and deleted before it can finish.

**Comment by willmot** [comment]
We also need to look into the consequences of this warning from [php.net](http://php.net/manual/en/function.flock.php)

> On some operating systems flock() is implemented at the process level. When using a multithreaded server API like ISAPI you may not be able to rely on flock() to protect files against other PHP scripts running in parallel threads of the same server instance!

Specifically, what does this mean for locking files in multi-server setups?

**Comment by willmot** [comment]
> Specifically, what does this mean for locking files in multi-server setups?

Here's my understanding:

1. `flock` file locking is process-based. That means the locks are only respected by the current PHP process. Multiple threads from the same process work correctly.
2. In multi-server environments it'll depend on whether the shared filesystem supports `flock` or not. 
    - NFS support `flock`.
    - S3 Uploads won't.

- [ ] We should provide a Constant to bypass it. That way, users who's FS doesn't support locking have a way forward.

**Comment by willmot** [comment]
I think this one is ready to go!

We'll want to do a beta before it's released though so let's leave it in it's own branch for now.

**Comment by katmoody** [comment]
@willmot - do we want to test this on folks whose sites aren't consistently running backups because of the spinning wheel of doom? Is it ready for beta testing in that case?

**Comment by willmot** [comment]
@katmoody yep exactly, once Dasha has reviewed I'll create a beta release zip that we can share with people.

**Comment by dashaluna** [comment]
@willmot I've looked at the code, however, my server knowledge isn't great. So, I don't understand what's going on :( Would you mind assigning the review of this to someone else? Apologies.

File: classes/backup/class-backup-status.php [hunk 1]
```diff
@@ -2,44 +2,209 @@
...
 
+	/**
+	 * The filename for the backup file we are the tracking status of.
+	 *
+	 * @var string
+	 */
 	private $filename = '';
 
+	/**
+	 * [$lock_handler description]
```
Comment (rmccue) [inline classes/backup/class-backup-status.php]: This needs an actual description.

File: classes/backup/class-backup-status.php [hunk 2]
```diff
@@ -2,44 +2,209 @@
...
+	 *
+	 * @var LockHandler
+	 */
+	private $lock_handler = '';
+
+	private $callback;
```
Comment (rmccue) [inline classes/backup/class-backup-status.php]: Needs phpDoc

File: classes/backup/class-backup-status.php [hunk 3]
```diff
@@ -2,44 +2,209 @@
...
+	 */
 	public function is_started() {
 		return (bool) file_exists( $this->get_status_filepath() );
 	}
 
-	public function finish() {
-		// Delete the backup running file
-		if ( file_exists( $this->get_status_filepath() ) ) {
-			unlink( $this->get_status_filepath() );
+	public function is_running() {
```
Comment (rmccue) [inline classes/backup/class-backup-status.php]: phpDoc

File: classes/backup/class-backup-status.php [hunk 4]
```diff
@@ -2,44 +2,209 @@
...
+
+			// If we're in the same thread then we know we must be running if the running file exists.
+			if ( is_a( $this->lock_handler, 'LockHandler' ) ) {
+				return $this->is_started();
+			}
+
+			$lock_handler = new LockHandler( basename( $this->get_status_filepath() ), Path::get_path() );
+
+			return ! $lock_handler->lock();
+		}
```
Comment (rmccue) [inline classes/backup/class-backup-status.php]: This conditional should be flipped, and returned early instead.

File: classes/backup/class-backup-status.php [hunk 5]
```diff
@@ -2,44 +2,209 @@
...
+	 */
+	public function catch_fatals() {
+
+		$error = error_get_last();
+
+		if ( empty( $error ) ) {
+			return;
+		}
+
+		if ( ! isset( $error['type'] ) || ! defined( 'E_ERROR' ) || E_ERROR !== $error['type'] ) {
```
Comment (rmccue) [inline classes/backup/class-backup-status.php]: `E_ERROR` is always defined; should also catch `E_USER_ERROR` here.

File: classes/backup/class-backup-status.php [hunk 6]
```diff
@@ -2,44 +2,209 @@
...
+
+		if ( $this->is_started() ) {
+			$status = json_decode( file_get_contents( $this->get_status_filepath() ) );
+
+			if ( ! empty( $status->filename ) ) {
+				$this->filename = $status->filename;
+			}
+		}
+
+		return $this->filename;
```
Comment (rmccue) [inline classes/backup/class-backup-status.php]: Seems a bit weird to always return this even if we haven't actually gotten a filename from the file

File: classes/backup/class-backup-status.php [hunk 7]
```diff
@@ -105,15 +272,18 @@ public function get_start_time() {
...
-	 * Get the path to the backup running file that stores the running backup status
+	 * Get the path to the backup running file that stores the running backup status.
 	 *
 	 * @return string
 	 */
 	public function get_status_filepath() {
 		return Path::get_path() . '/.backup-' . $this->id . '-running';
 	}
+
+	public function set_status_callback( $callback ) {
```
Comment (rmccue) [inline classes/backup/class-backup-status.php]: phpDoc

**Comment by willmot** [comment]
@sambulance  @shadyvb I've struggled to find time to fixup Ryan's comments above. If either of you want to take a bash it would be great to get this in.

**Pull request closed without merging**